### ISSUE TITLE
Fitting Bayesian Networks in parallel

### ISSUE THREAD

**Issue by UndeadKernel** [jmschrei/pomegranate#258]
Bayesian networks cannot be fit using the parallel functionality provided.
Is it viable to add parallelization to the summarization process done in the "summarize" function?
It would be beneficial, while this functionality is not implemented, to inform the user with appropriate error messages whenever parallelization functions are used to fit Bayesian networks.

**Comment by jmschrei** [comment]
Can you share the code that you tried to use that does not work?

**Comment by UndeadKernel** [comment]
This is a toy problem that tries to fit data samples to a bayesian network:
You can [find here](https://github.com/jmschrei/pomegranate/files/1000973/samples.txt) a .txt file with 2,000 samples formatted as JSON.
The following code first builds a Bayesian Network. Afterwards, it loads the samples from a file (make sure that the samples are in the same directory as the script) and then tries to use the parallel functionality provided.

Let me know if you are able to run this sample script and if you need help in any other way.

```
from pomegranate import *
import json

A = DiscreteDistribution({'zero': 0, 'one': 0})
B = DiscreteDistribution({'zero': 0, 'one': 0})
C = ConditionalProbabilityTable(
    [['zero','zero','zero',0],
     ['zero','zero','one',0],
     ['zero','one','zero',0],
     ['zero','one','one',0],
     ['one','zero','zero',0],
     ['one','zero','one',0],
     ['one','one','zero',0],
     ['one','one','one',0]], [A, B])
D = ConditionalProbabilityTable(
    [['zero','zero',0],
     ['zero','one',0],
     ['one','zero',0],
     ['one','one',0]], [C])

As = State(A, name="A")
Bs = State(B, name="B")
Cs = State(C, name="C")
Ds = State(D, name="D")

model = BayesianNetwork("Toy Problem")

model.add_states(As, Bs, Cs, Ds)

model.add_transition(As, Cs)
model.add_transition(Bs, Cs)
model.add_transition(Cs, Ds)
model.bake()

# Load the samples
samples = None
with open('samples.txt', 'r') as f:
    samples = json.load(f)
# Transform the data to pomegranate's format
pome_samples = [[s['A'], s['B'], s['C'], s['D']] for s in samples]

# Try to fit the model
fit(model, pome_samples, n_jobs=4)
```

**Comment by jmschrei** [comment]
Thanks, I can reproduce the error. It may be possible that fitting Bayesian networks without parallelization may be fast enough for your purposes, as it's just a simple MLE update. I'll let you know when I fix the issue.

**Comment by UndeadKernel** [comment]
Parallelization might still be useful when the network is very large. For example, I was fitting networks that had about 140 nodes (with some nodes having up to five parents) and 20 million data samples. Fitting the model would take many hours. As MLE can be calculated following a map-reduce approach, parallelization would definitely speed things up.

Cheers for the great work already done in this library, by the way.

**Comment by jmschrei** [comment]
Agreed. Parallelism is always good, especially as models get large. Getting rid of the wrappers and adding in better model-specific parallelism is one of my goals that shouldn't be too difficult for Bayesian networks. 

I'm glad that you appreciate it! Please let me know if you have any other issues or comments, as I'd prefer my development follow user need as opposed to personal whimsy.

**Comment by TuhinSharma** [comment]
I have a predefined network structure and based on the co-occurence of various nodes I want to construct the Bayesian Network. 

So I have defined a function which creates a random matrix with shape (record_count,node_count) having only 1s and 0s, a function which creates a dependency tuple (network structure ) for those node_count number of nodes, and a function which returns the list of node names for those node_count number of nodes. I have used **BayesianNetwork.from_structure()** for constructing and training the network. Can I do it using parallelism? The code is as follows which works fine without parallelism:-

```
from pomegranate import *
import numpy as np
import random as rd
import time
import pickle

def generate_matrix_with_shape( m, n):
    x = np.random.randint(10, size=(m*n-n))
    y = np.array([1 if i<2 else 0 for i in x])
    z = y.reshape((m-1,n))
    result = np.vstack((z,np.ones((1,n))))
    return result


def generate_shallow_pomegranate_structure(n):
    l = list()
    
    if n>2:
        for i in range(n-3):
            x = tuple(np.array(rd.sample(range(i+1,n), 3)))
            l.append(x)
        l.append(())
        l.append(())
        l.append(())
    elif n==2:
        l.append((1,))
        l.append(())
    else:
        raise BaseException("enter n>=2")
        return None
    return tuple(l)

def generate_wide_pomegranate_structure(n):
    l=list()
    m=30
    if n>m:
        for i in range(n-m):
            rand_num = n-m + np.random.randint(m-5)
            x = tuple(np.array([rand_num]))
            l.append(x)
        for j in range(m-5):
            rand_num = n-5 + np.random.randint(5)
            x = tuple(np.array([rand_num]))
            l.append(x)
        l.append(())
        l.append(())
        l.append(())
        l.append(())
        l.append(())
            
    elif n==2:
        l.append((1,))
        l.append(())
    else:
        raise BaseException("enter node_count>=",m)
    return tuple(l) 


def generate_state_names(n):
    l = list()
    for i in range(n):
        l.append(str(i))
    return l
        
    
    

record_count = 5
node_count = 40


matrix = generate_matrix_with_shape(record_count, node_count)

print("matrix generation successful.\n")
structure = generate_wide_pomegranate_structure(node_count)

print("PGM structure generated.\n")
state_names = generate_state_names(node_count)

print("state_names generated.\n")

tic = time.time()

print("Model training started.\n")

model = BayesianNetwork.from_structure(matrix, structure=structure, state_names=state_names)

print ("Model training done.\n")

training_time = time.time() - tic

print("training time : ",training_time)
print()
```

**Comment by jmschrei** [comment]
Unfortunately, for the same reason that `fit` does not currently allow parallelism, `from_structure` does not either. Basically, `from_structure` is just a wrapper for `fit` where the network is first initialized according to the structure you defined. Sorry about the inconvenience, I'll update this thread when I add it in.

**Comment by TuhinSharma** [comment]
I have failed to use ```predict_proba()``` as well using parallelism. The docstring says that it requires  np.ndarray() but if I am not wrong, ```BayesianNetwork. predict_proba``` requires a ```dict()```. Am I doing something wrong? should I create a new isseue to address this?

**Comment by jmschrei** [comment]
Right now `predict_proba` takes in a single dict corresponding to the observed features of a single sample. I'm trying to determine the best way to allow it to take in an array of samples, but run into the problem where it has to return a probability distribution for each missing sample instead of a single value. I don't think it is currently parallelized, but you should be able to do it yourself using the following:

```python
from pomegranate import BayesianNetwork
from joblib import Parallel, delayed

def parallel_predict(filename, X):
    model = BayesianNetwork.from_json(filename)
    return map(model.predict_proba, X)

filename = 'bayesian_network.json'
n_jobs = .. however many jobs ..

model = ... however you create your model .. 
model.to_json(filename)

X = [ ... an array of dicts, where each dict is {'observed feature': 'observed value'} ]
X_imputed = reduce(list.__add__, Parallel(n_jobs=n_jobs)( delayed(parallel_predict)(filename, X[i::n_jobs] for i in range(n_jobs) ))
```

You may need to tweak that script a little bit more, but roughly it should work. Let me know if it doesn't.

**Comment by TuhinSharma** [comment]
I dont know how to use ```joblib```. But here is the script what I tried: 
```
from pomegranate import *
import numpy as np
import random as rd
import time
import pickle
from joblib import Parallel, delayed

def generate_matrix_with_shape( m, n):
    x = np.random.randint(10, size=(n*(m-2)))
    y = np.array([1 if i<2 else 0 for i in x])
    z = y.reshape((m-2,n))
    result = np.vstack((z,np.ones((1,n))))
    result = np.vstack((result,np.zeros((1,n))))
    return result


def generate_shallow_pomegranate_structure(n):
    l = list()
    
    if n>2:
        for i in range(n-3):
            x = tuple(np.array(rd.sample(range(i+1,n), 3)))
            l.append(x)
        l.append(())
        l.append(())
        l.append(())
    elif n==2:
        l.append((1,))
        l.append(())
    else:
        raise BaseException("enter n>=2")
        return None
    return tuple(l)

def generate_wide_pomegranate_structure(n):
    l=list()
    m=30
    if n>m:
        for i in range(n-m):
            rand_num = n-m + np.random.randint(m-5)
            x = tuple(np.array([rand_num]))
            l.append(x)
        for j in range(m-5):
            rand_num = n-5 + np.random.randint(5)
            x = tuple(np.array([rand_num]))
            l.append(x)
        l.append(())
        l.append(())
        l.append(())
        l.append(())
        l.append(())
            
    elif n==2:
        l.append((1,))
        l.append(())
    else:
        raise BaseException("enter node_count>=",m)
    return tuple(l) 


def generate_state_names(n):
    l = list()
    for i in range(n):
        l.append(str(i))
    return l
        
    
    

record_count = 5
node_count = 40



matrix = generate_matrix_with_shape(record_count, node_count)

print("matrix generation successful.\n")
structure = generate_wide_pomegranate_structure(node_count)

print("PGM structure generated.\n")
state_names = generate_state_names(node_count)

print("state_names generated.\n")

tic = time.time()

print("Model training started.\n")

model = BayesianNetwork.from_structure(matrix, structure=structure, state_names=state_names)

print ("Model training done.\n")

training_time = time.time() - tic

print("training time : ",training_time)
print()


def parallel_predict(model, X):
    return map(model.predict_proba, X)

n_jobs = 5


X = [ {"0":1},{"1":1} ]
n = len(X)

X_imputed = reduce(list.__add__, Parallel(n_jobs=n_jobs)( delayed(parallel_predict)(model, X[n/n_jobs*i:n/n_jobs*(i+1)] for i in range(n_jobs)) ))
```

I could not run this. Please help me. The matrix used for training is binary matrix with 0s and 1s.

**Comment by jmschrei** [comment]
There was a small bug in my code. There is also a small bug in pickling the BayesianNetwork object so you'll need to save it to a json and load it up in each process, which is pretty fast (~0.02 seconds for your model).

Here is the code I used (I only modified things under `def parallel_predict`:

```python
from pomegranate import *
import numpy as np
import random as rd
import time
import pickle
from joblib import Parallel, delayed

def generate_matrix_with_shape( m, n):
    x = np.random.randint(10, size=(n*(m-2)))
    y = np.array([1 if i<2 else 0 for i in x])
    z = y.reshape((m-2,n))
    result = np.vstack((z,np.ones((1,n))))
    result = np.vstack((result,np.zeros((1,n))))
    return result


def generate_shallow_pomegranate_structure(n):
    l = list()
    
    if n>2:
        for i in range(n-3):
            x = tuple(np.array(rd.sample(range(i+1,n), 3)))
            l.append(x)
        l.append(())
        l.append(())
        l.append(())
    elif n==2:
        l.append((1,))
        l.append(())
    else:
        raise BaseException("enter n>=2")
        return None
    return tuple(l)

def generate_wide_pomegranate_structure(n):
    l=list()
    m=30
    if n>m:
        for i in range(n-m):
            rand_num = n-m + np.random.randint(m-5)
            x = tuple(np.array([rand_num]))
            l.append(x)
        for j in range(m-5):
            rand_num = n-5 + np.random.randint(5)
            x = tuple(np.array([rand_num]))
            l.append(x)
        l.append(())
        l.append(())
        l.append(())
        l.append(())
        l.append(())
            
    elif n==2:
        l.append((1,))
        l.append(())
    else:
        raise BaseException("enter node_count>=",m)
    return tuple(l) 


def generate_state_names(n):
    l = list()
    for i in range(n):
        l.append(str(i))
    return l
        
    
    

record_count = 5
node_count = 40



matrix = generate_matrix_with_shape(record_count, node_count)

print("matrix generation successful.\n")
structure = generate_wide_pomegranate_structure(node_count)

print("PGM structure generated.\n")
state_names = generate_state_names(node_count)

print("state_names generated.\n")

tic = time.time()

print("Model training started.\n")

model = BayesianNetwork.from_structure(matrix, structure=structure, state_names=state_names)

print ("Model training done.\n")

training_time = time.time() - tic

print("training time : ",training_time)
print()


def parallel_predict(filename, X):
    model = BayesianNetwork.from_json(filename)
    return map(model.predict_proba, X)

X = [{"0":1},{"1":1},{"0":1},{"1":1},{"0":1}]*100

n = len(X)

with open("test.json", "w") as outfile:
    outfile.write(model.to_json())

for n_jobs in 1, 2, 4, 8, 16:
    tic = time.time()
    X_imputed = reduce(list.__add__, Parallel(n_jobs=n_jobs)( delayed(parallel_predict)("test.json", X[n/n_jobs*i:n/n_jobs*(i+1)]) for i in range(n_jobs+1) ))
    print "With {} jobs, takes {:4.4}s to run on {} samples".format(n_jobs, time.time() - tic, len(X_imputed))
```

and the results that I got:

```
[jmschr@n026 pomegranate]$ python bn_proba_test.py
matrix generation successful.

PGM structure generated.

state_names generated.

Model training started.

Model training done.

('training time : ', 0.00852513313293457)
()
With 1 jobs, takes 8.703s to run on 500 samples
With 2 jobs, takes 5.292s to run on 500 samples
With 4 jobs, takes 2.659s to run on 500 samples
With 8 jobs, takes 1.448s to run on 500 samples
With 16 jobs, takes 0.9936s to run on 500 samples
```

**Comment by TuhinSharma** [comment]
Thanks. It worked!! By the way I have another question. Is there a way to make the Bayesian Network learn incrementally based on current implementation? To elaborate, I have a trained model (100 nodes) based on 5000 records (numpy array with shape (5000,100)). Tomorrow I want to include another array with shape (4000,100). How can I fit the already trained model on this data so that it would effectively get trained on the whole dataset, i.e of shape (9000,100)?

**Comment by jmschrei** [comment]
Howdy. You should open a new issue for each question so that others who may have the same question can easily search for them in the future (and limit an issue to a single topic).

One way you can do that is to simply fit it to the entire dataset the second day. It doesn't seem like that much data. Another way you can do it is to do `model.summarize(day1)` then `model.summarize(day2)`, then `model.from_summaries()` to exploit the out-of-core API. When you do `from_summaries` it clears the stored summaries. I am considering adding in an option to make it not clear those, but it's not there yet.

**Comment by jmschrei** [comment]
I have since added n_jobs to both `predict_proba` and `fit`. Let me know how it goes.

**Comment by marchezinixd** [comment]
Is n_jobs working? I tried to put 1,2,3,4 when fitting, always the same time and result.
Actually setting n_jobs does not affect neither structure learning nor fitting. I'm handling a subset of my dataset that have 100 variables and 5 million rows, but it will increase to 5000 variables with 5 million rows.  If this is not parallel it will not be possible for me to run.

**Comment by AdirthaBorgohain** [comment]
Sorry for recommenting in this issue. I can see n_jobs in ```predict_proba()``` and ```fit()```. But is there an way to use n_jobs using ```from_structure()```as well?

**Comment by jmschrei** [comment]
There currently isn't. The general structure finding process we use is a greedy algorithm, and while each iteration of the algorithm can be parallelized, it doesn't provide significant speed benefits.

**Comment by Piyushbalwani** [comment]
Sorry, I am asking same question, but I didn't understand solution for this issue.
If number of samples are in thousands, so matrix will be out of RAM for passing as argument to `from_structure`. And as you suggested, we can't use `n_jobs `(as not supported for `from_structure `) or `summarize` (as I only have initial structure and data samples, but not Conditional probability table). Can you please suggest how can I use from_structure for large number of samples ? Can `joblib `help here, but I am not getting how can I twist `joblib `for `from_structure`? Thanks!!!
`model = BayesianNetwork.from_structure(matrix, structure=structure, state_names=state_names)
`

**Comment by jmschrei** [comment]
What does your structure look like? I don't think that you'll run out of RAM because of the number of samples. It is much more likely that you are specifying a table that is much too large. Remember that tables takes memory proportional to the product of the number of keys in each column. If your data is binary and one variable has 15 parents, the table will have 2**(15+1) entries. I don't think parallelizing anything will overcome memory issues.

**Comment by Piyushbalwani** [comment]
Thanks for explanation. My Bayesian structure is like bi-partite graph, where edges are directed from first level nodes (non-binary 32 discrete value) to the second level node. Each second level node has at-least 50 parents from first level. So, my goal was to learn structure and parameters. I tried to learn structure with `from_samples`, but it is taking indefinite time with `exact `algorithm. And now trying to learn parameters, if given structure, but seems memory is causing the issue.

**Comment by jmschrei** [comment]
Gotcha. Yes, that seems likely to cause a memory issue because you need to make a table with at least 32 ** 50 entries which is a lot. I'd recommend trying to find a way to compress the value space and have fewer parents per node, if possible.

**Comment by Piyushbalwani** [comment]
Thanks for your great help and rescue!!! So, even if we reduce it to 4 parents (32 values) per node (8 values), table would have around 8*(32^4)~8million entries. So, to learn this many values, as per my opinion,  at-least, we need to consider 80 million sample points. Will `summarize` help here in learning parameters with so many samples points, like batch mode?
`model = BayesianNetwork.from_structure(X, structure=s, state_names=s, keys=k)
`
(It seems key is necessary to pass for further summarize)
```
model.summarize(X)
while(njob!=1):
 X=generate_dataset()
 model.summarize(X)
 njob=njob-1
model.from_summaries()
print(model)
```
Jacob, can you please tell is this correct code and approach for learning parameters with large sample points?

**Comment by Piyushbalwani** [comment]
Hello Jacob,

1. Do you know any way or algorithm to compress the parents so that Bayesian structure reduced. I am thinking if we can append some nodes in between parent and leaf node (like 64 parent to 1 leaf node converted into 64 parent to 16 node, 16 node to 4 node, 4 node to 1leaf ), so that conditional table variables reduced.

2. Can I use `summarize ` and `from_summaries `for `from_strucuture `  model for out-of-core parameter learning?
```
model = BayesianNetwork.from_structure
model.summarize(X)
while(njob!=1):
 X=generate_dataset()
 model.summarize(X)
 njob -= 1
model.from_summaries()
```

Thanks and Regards
Piyush

**Comment by jmschrei** [comment]
> Thanks for your great help and rescue!!! So, even if we reduce it to 4 parents (32 values) per node (8 values), table would have around 8*(32^4)~8million entries. So, to learn this many values, as per my opinion, at-least, we need to consider 80 million sample points. Will `summarize` help here in learning parameters with so many samples points, like batch mode?
> `model = BayesianNetwork.from_structure(X, structure=s, state_names=s, keys=k) `
> (It seems key is necessary to pass for further summarize)
> 
> ```
> model.summarize(X)
> while(njob!=1):
>  X=generate_dataset()
>  model.summarize(X)
>  njob=njob-1
> model.from_summaries()
> print(model)
> ```
> 
> Jacob, can you please tell is this correct code and approach for learning parameters with large sample points?

Yes, that approach looks like it will work. Keep in mind that the structure will be based off the first batch but the parameters will be based on the entire data set. It is not feasible to learn the structure of a Bayesian network over a stream of data.


> Hello Jacob,
> 
> 1. Do you know any way or algorithm to compress the parents so that Bayesian structure reduced. I am thinking if we can append some nodes in between parent and leaf node (like 64 parent to 1 leaf node converted into 64 parent to 16 node, 16 node to 4 node, 4 node to 1leaf ), so that conditional table variables reduced.
> 2. Can I use `summarize ` and `from_summaries `for `from_strucuture `  model for out-of-core parameter learning?
> 
> ```
> model = BayesianNetwork.from_structure(X, structure=s, state_names=s, keys=k)
> model.summarize(X)
> while(njob!=1):
>  X=generate_dataset()
>  model.summarize(X)
>  njob -= 1
> model.from_summaries()
> ```
> 
> Thanks and Regards
> Piyush

1. If you know this type of structure in advance (which nodes should be the great grandparents, the parents, and the children, but you don't know the connections between them) you should consider using a constraint graph. There is a tutorial on how to use them in the tutorial folder of the repo. If you don't know this info in advance, in principle it is possible to define hidden nodes that are not directly observed in the data set but can be used to compress the data. Unfortunately, hidden nodes are not currently implemented in pomegranate. If the precise value of the parents is not important you could try binning it more, e.g. using only the first three most likely values and than an "other" indicator. If the values are discrete values but still ordinal (such as pixel values) you could bin the range into four quartiles. That would get the number of parameters down from 32 ** 50 to 4 ** 50. 

2. Yes. That will work when your data set doesn't fit in memory but it won't fix the problem of your model being too big, if that is the problem.
